### PR TITLE
Restructure and modernize tour of C#

### LINKED ISSUE #9576: Push/Pop Example
It's been about a year since I went through the C# docs, so I'm probably missing something mentioned somewhere.  Even after loading the example into VS, I am unsure of the functional nature of the class type objects.  It is hard for me to explain concisely, so I will walk through my understanding (combined example with the data supplied): Begin in main & create Stack obj 's' which has the field 'top' set to null.  Invoke the Push method of s with an int arg of 1 supplied to satisfy the object 'data' parameter.  This in turn assigns a new obj of class type 'Entry' to s's 'top' field, which means the field now represents the values of Entry's instance constructor's 'next' and 'data' fields (which themselves are set to the value of Entry's instance variables of type Entry and type object).  (I'm ignoring the next two method calls to Push) We then have the call to Pop which is also a method of s's and do a conditional check to see if 'top' is null.  It isn't so the field of type object, 'result', is assigned the value of 'top.data'.  Then my brain explodes bc I'm not sure how the assignment of 'top' to 'top.next' effectively discards the prior data and moves it to the next referenced storage location.

It may be that I'm missing something that isn't shown in the locals or autos windows.  Is this the way the array methods Push & Pop are represented internally?  Thank you.

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: 20935a57-1ba9-a5b8-0a30-cc68dac2e4e1
* Version Independent ID: 7c81da48-bf33-8c1a-f9fe-d02e056b2a07
* Content: [C# Program Structure - A Tour of the C# Language](https://docs.microsoft.com/en-us/dotnet/csharp/tour-of-csharp/program-structure)
* Content Source: [docs/csharp/tour-of-csharp/program-structure.md](https://github.com/dotnet/docs/blob/master/docs/csharp/tour-of-csharp/program-structure.md)
* Product: **dotnet-csharp**
* GitHub Login: @BillWagner
* Microsoft Alias: **wiwagn**

### LINKED ISSUE #13640: Events vague
The events section is rather vague, and doesn't contain any links to more information. For example, the EventHandler type itself is never described, the ?. operator isn't explained in this series at all IIRC, and the EventArgs type isn't described. If nothing else, a link to more information would be handy.

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: 8181f41d-b6ce-ad66-084e-8d2b514b87e2
* Version Independent ID: 494c4344-b752-4f86-5b66-1ef4b81c1e75
* Content: [Classes and Objects in C# - A tour of the C# Language](https://docs.microsoft.com/en-us/dotnet/csharp/tour-of-csharp/classes-and-objects#feedback)
* Content Source: [docs/csharp/tour-of-csharp/classes-and-objects.md](https://github.com/dotnet/docs/blob/master/docs/csharp/tour-of-csharp/classes-and-objects.md)
* Product: **dotnet-csharp**
* GitHub Login: @BillWagner
* Microsoft Alias: **wiwagn**

### LINKED ISSUE #16176: What's "a compatible array type"?
I think that'd be good if there's some explanation somewhere in C# docs.

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: b9770af4-de82-7034-a56b-c68a6b5c55e5
* Version Independent ID: 6bd64fcc-5c84-e541-12f4-0cdcfe4f5dd2
* Content: [C# Types and Variables - A tour of the C# language](https://docs.microsoft.com/en-us/dotnet/csharp/tour-of-csharp/types-and-variables)
* Content Source: [docs/csharp/tour-of-csharp/types-and-variables.md](https://github.com/dotnet/docs/blob/master/docs/csharp/tour-of-csharp/types-and-variables.md)
* Product: **dotnet-csharp**
* Technology: **csharp-get-started**
* GitHub Login: @BillWagner
* Microsoft Alias: **wiwagn**

### LINKED ISSUE #18182: Introduction
Nice. Can be little bit pictorial if possible?

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: 14618620-83fc-fd68-93a9-03626c76d680
* Version Independent ID: ea7fbbaf-fd41-75cb-28fb-a2f36c8d3967
* Content: [A Tour of C# - C# Guide](https://docs.microsoft.com/en-us/dotnet/csharp/tour-of-csharp/#feedback)
* Content Source: [docs/csharp/tour-of-csharp/index.md](https://github.com/dotnet/docs/blob/master/docs/csharp/tour-of-csharp/index.md)
* Product: **dotnet-csharp**
* Technology: **csharp-get-started**
* GitHub Login: @BillWagner
* Microsoft Alias: **wiwagn**

### LINKED ISSUE #18672: Explain variable naming constaints
I recommend adding a section that identifies variable naming constraints. Something along the lines of "You should avoid using these reserved words in a variable name: <bulleted list or link to doc here>. If you have to for interoperability with another language, prefix the variable name with `@`".

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: b9770af4-de82-7034-a56b-c68a6b5c55e5
* Version Independent ID: 6bd64fcc-5c84-e541-12f4-0cdcfe4f5dd2
* Content: [C# Types and Variables - A tour of the C# language](https://docs.microsoft.com/en-us/dotnet/csharp/tour-of-csharp/types-and-variables)
* Content Source: [docs/csharp/tour-of-csharp/types-and-variables.md](https://github.com/dotnet/docs/blob/master/docs/csharp/tour-of-csharp/types-and-variables.md)
* Product: **dotnet-csharp**
* Technology: **csharp-get-started**
* GitHub Login: @BillWagner
* Microsoft Alias: **wiwagn**

### PR DESCRIPTION
**Pull request by BillWagner**
Fixes #9576 
Fixes #13640 
Fixes #16176
Fixes #18182
Fixes #18672 

The Tour of C# has become very dated.

In this PR, I reviewed and updated each section:
- Emphasize .NET Core SDK tooling over `csc`
- Highlight major areas of C# added since 1.0 (Yes, this was that old)
- De-emphasize areas that aren't as common to use in modern C#.

[Internal review link](https://review.docs.microsoft.com/en-us/dotnet/csharp/tour-of-csharp/?branch=pr-en-us-18538)

### CODE REVIEW

File: docs/csharp/tour-of-csharp/features.md [hunk 1]
```diff
@@ -0,0 +1,118 @@
+---
+title: A Tour of C# - Major language areas
+description: New to C#? Learn the basics of the language.
+ms.date: 08/06/2020
+---
+# Major language areas
+
+## Arrays, collections, and LINQ
+
+C# and .NET provide many different collection types. `Arrays` have syntax defined by the language. General collection types are listed in the <xref:System.Collections.Generic?displayProperty=fullName> namespace. Specialized collections include <xref:System.Span%601?displayProperty=nameWithType> for accessing continuous memory on the stack frame, and <xref:System.Memory%601?displayProperty=nameWithType> for accessing continuous memory on the managed heap. All collections, including arrays, <xref:System.Span%601>, and <xref:System.Memory%601> share a unifying principle for iteration. You use the <xref:System.Collections.Generic.IEnumerable%601?displayProperty=nameWithType> interface. This unifying principle means that any of the collection types can be used with LINQ queries or other algorithms. You write methods using <xref:System.Collections.Generic.IEnumerable%601> and those algorithms work with any collection.
```
Comment (pkulikov) [inline docs/csharp/tour-of-csharp/features.md:10]: I don't think "Arrays" in the second sentence should be code fenced: there is no such type and it won't be localized, while it seems it should be.

File: docs/csharp/tour-of-csharp/program-building-blocks.md [hunk 2]
```diff
@@ -291,6 +233,67 @@ The garbage collector is allowed wide latitude in deciding when to collect objec
...
+
+Precedence and associativity can be controlled using parentheses. For example, `x + y * z` first multiplies `y` by `z` and then adds the result to `x`, but `(x + y) * z` first adds `x` and `y` and then multiplies the result by `z`.
+
+Most operators can be [*overloaded*](../language-reference/operators/operator-overloading.md). Operator overloading permits user-defined operator implementations to be specified for operations where one or both of the operands are of a user-defined class or struct type.
+
+C# provides a number of operators to perform [arithmetic](../language-reference/operators/arithmetic-operators.md), [logical](../language-reference/operators/boolean-logical-operators.md), [bitwise and shift](../language-reference/operators/bitwise-and-shift-operators.md) operations and [equality](../language-reference/operators/equality-operators.md) and [order](../language-reference/operators/comparison-operators.md) comparisons.
+
+For the complete list of C# operators ordered by precedence level, see [C# operators](../language-reference/operators/index.md).
```
Comment (pkulikov) [inline docs/csharp/tour-of-csharp/program-building-blocks.md:253]: This section duplicates a lot the referenced article. Is that by design?

File: docs/csharp/tour-of-csharp/arrays.md [hunk 3]
```diff
@@ -1,43 +0,0 @@
----
```
Comment (pkulikov) [inline docs/csharp/tour-of-csharp/arrays.md:1]: Are no updates in the redirection file by design?

File: docs/csharp/tour-of-csharp/features.md [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+
+:::code language="csharp" source="./snippets/shared/Features.cs" ID="InitializeGenerated":::
+
+The `foreach` statement can be used to enumerate the elements of any collection. The following code enumerates the array from the preceding example:
+
+:::code language="csharp" source="./snippets/shared/Features.cs" ID="EnumerateArray":::
+
+The `foreach` statement uses the <xref:System.Collections.Generic.IEnumerable%601> interface, so can work with any collection.
+
+## string interpolation
```
Comment (pkulikov) [inline docs/csharp/tour-of-csharp/features.md:51]:
```suggestion
## String interpolation
```

File: docs/csharp/tour-of-csharp/features.md [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+
+C# [***string interpolation***](../language-reference/tokens/interpolated.md) enables you to format strings by define expressions whose results are placed in a format string. For example, the following example prints the temperature on a given day from a set of weather data:
+
+:::code language="csharp" source="./snippets/shared/Features.cs" ID="StringInterpolation":::
+
+An interpolated string is declared using the `$` token. String interpolation evaluates the expressions between `{` and `}`, then converts the result to a `string`, and replaces the text between the brackets with the string result of the expression. The `:` in the first expression, `{weatherData.Data:MM-DD-YYYY}` specifies the *format string*. In the preceding example, it specifies that the date should be printed in "MM-DD-YYYY" format.
+
+## Pattern Matching
```
Comment (pkulikov) [inline docs/csharp/tour-of-csharp/features.md:59]:
```suggestion
## Pattern matching
```

File: docs/csharp/tour-of-csharp/features.md [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+
+:::code language="csharp" source="./snippets/shared/Features.cs" ID="UseDelegate":::
+
+A delegate doesn't know or care about the class of the method it references. All that matters is that the referenced method has the same parameters and return type as the delegate.
+
+## async / await
+
+C# supports asynchronous programs with two keywords: `async` and `await`. You add the `async` modifier to a method declaration to declare the method is asynchronous. The `aawit` operator tells the compiler that to asynchronously await for a result to finish. Control is returned to the caller, and the method returns a structure that manages the state of the asynchronous work. The structure is typically a <xref:System.Threading.Tasks.Task%601?displayProperty=nameWithType>, but can be any type that supports the awaiter pattern. These features enable you to write code that reads as its synchronous counterpart, but executes asynchronously. For example, the following code downloads the home page for [Microsoft docs](https://docs.microsoft.com):
```
Comment (pkulikov) [inline docs/csharp/tour-of-csharp/features.md:83]:
```suggestion
C# supports asynchronous programs with two keywords: `async` and `await`. You add the `async` modifier to a method declaration to declare the method is asynchronous. The `await` operator tells the compiler that to asynchronously await for a result to finish. Control is returned to the caller, and the method returns a structure that manages the state of the asynchronous work. The structure is typically a <xref:System.Threading.Tasks.Task%601?displayProperty=nameWithType>, but can be any type that supports the awaiter pattern. These features enable you to write code that reads as its synchronous counterpart, but executes asynchronously. For example, the following code downloads the home page for [Microsoft docs](https://docs.microsoft.com):
```

File: docs/csharp/tour-of-csharp/program-building-blocks.md [hunk 7]
```diff
@@ -291,6 +233,67 @@ The garbage collector is allowed wide latitude in deciding when to collect objec
...
+
+## Statements
```
Comment (pkulikov) [inline docs/csharp/tour-of-csharp/program-building-blocks.md:255]: Again, it duplicates a lot the existing documentation: https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/statements-expressions-operators/statements

I'm not sure it makes the tour approachable. To me, it goes into too many details.

File: docs/csharp/tour-of-csharp/program-building-blocks.md [hunk 8]
```diff
@@ -291,6 +233,67 @@ The garbage collector is allowed wide latitude in deciding when to collect objec
...
+* `continue` statement.
+* `goto` statement.
+* `return` statement.
+* `yield` statement.
+* `throw` statements and `try` statements.
+* `checked` and `unchecked` statements.
+* `lock` statement.
+* `using` statement.
+
+When a method body is a single expression, the method can be defined using a compact expression format, as shown in the following example:
```
Comment (pkulikov) [inline docs/csharp/tour-of-csharp/program-building-blocks.md:291]: This fits more the section about methods (how to concisely define methods), not the one about statements.

File: docs/csharp/tour-of-csharp/types.md [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+
+An implicit conversion exists from a class type to any of its base class types. A variable of a class type can reference an instance of that class or an instance of any derived class. For example, given the previous class declarations, a variable of type `Point` can reference either a `Point` or a `Point3D`:
+
+:::code language="csharp" source="./snippets/shared/Types.cs" ID="ImplicitCastToBase":::
+
+## Structs
+
+Classes define types that support inheritance and polymorphism. They enable you to create sophisticated behaviors based on hierarchies of derived classes. By contrast, [***struct***](../language-reference/builtin-types/value-types.md) types are simpler types whose primary purpose is to store data values. Structs can't declare a base type; they implicitly derive from <xref:System.ValueType?displayProperty=nameWithType>. You can't derive other `struct` types from a `struct` type. They're implicitly sealed.
```
Comment (pkulikov) [inline docs/csharp/tour-of-csharp/types.md:58]:
```suggestion
Classes define types that support inheritance and polymorphism. They enable you to create sophisticated behaviors based on hierarchies of derived classes. By contrast, [***struct***](../language-reference/builtin-types/struct.md) types are simpler types whose primary purpose is to store data values. Structs can't declare a base type; they implicitly derive from <xref:System.ValueType?displayProperty=nameWithType>. You can't derive other `struct` types from a `struct` type. They're implicitly sealed.
```

Comment (BillWagner) [reply docs/csharp/tour-of-csharp/types.md:58]: Good change. (Interesting in that I was using the TOC filter, and it didn't bring this article up. I'll check again)

Comment (pkulikov) [reply docs/csharp/tour-of-csharp/types.md]: @BillWagner I've just checked the TOC filter with the "struct" entry. My observations:
- A lot of articles about con**struct**ors are brought up
- Suggestions look to be sorted by order in TOC. The language reference node is almost the last one (after the programming guide one).
- That makes the "Structure types" article suggestion to be the second from the list bottom. You should scroll down through the brought up suggestions to notice it.

Searching for "structure" in the TOC filter is better.

File: docs/csharp/tour-of-csharp/snippets/shared/ClassesObjects.cs [hunk 10]
```diff
@@ -0,0 +1,453 @@
...
+        }
+        static void F<T>(T x)
+        {
+            Console.WriteLine("F<T>(T)");
+        }
+        static void F(double x, double y)
+        {
+            Console.WriteLine("F(double, double)");
+        }
+        public static void UsageExample()
```
Comment (IEvangelist) [inline docs/csharp/tour-of-csharp/snippets/shared/ClassesObjects.cs:165]:
```suggestion
        static void F() => Console.WriteLine("F()");
        static void F(object x) => Console.WriteLine("F(object)");
        static void F(int x) => Console.WriteLine("F(int)");
        static void F(double x) => Console.WriteLine("F(double)");
        static void F<T>(T x) => Console.WriteLine("F<T>(T)");            
        static void F(double x, double y) => Console.WriteLine("F(double, double)");
        
        public static void UsageExample()
```

File: docs/csharp/tour-of-csharp/arrays.md (hunk 3 above)
Comment (BillWagner) [reply docs/csharp/tour-of-csharp/arrays.md:1]: No, thanks for the reminder.

File: docs/csharp/tour-of-csharp/snippets/shared/ClassesObjects.cs [hunk 11]
```diff
@@ -0,0 +1,453 @@
...
+
+        // Event
+        public event EventHandler Changed;
+
+        // Operators
+        public static bool operator ==(MyList<T> a, MyList<T> b) =>
+            Equals(a, b);
+
+        public static bool operator !=(MyList<T> a, MyList<T> b) =>
+            !Equals(a, b);
```
Comment (IEvangelist) [inline docs/csharp/tour-of-csharp/snippets/shared/ClassesObjects.cs:264]:
```suggestion
        const int DefaultCapacity = 4;

        T[] _items;
        int _count;

        public MyList(int capacity = DefaultCapacity)
        {
            _items = new T[capacity];
        }

        public int Count => _count;

        public int Capacity
        {
            get =>  _items.Length;
            set
            {
                if (value < _count) value = _count;
                if (value != _items.Length)
                {
                    T[] newItems = new T[value];
                    Array.Copy(_items, 0, newItems, 0, count);
                    _items = newItems;
                }
            }
        }

        public T this[int index]
        {
            get => _items[index];
            set
            {
                _items[index] = value;
                OnChanged();
            }
        }

        public void Add(T item)
        {
            if (_count == Capacity) Capacity = _count * 2;
            items[_count] = item;
            _count++;
            OnChanged();
        }
        protected virtual void OnChanged() =>
            Changed?.Invoke(this, EventArgs.Empty);

        public override bool Equals(object other) =>
            Equals(this, other as MyList<T>);

        static bool Equals(MyList<T> a, MyList<T> b)
        {
            if (Object.ReferenceEquals(a, null)) return Object.ReferenceEquals(b, null);
            if (Object.ReferenceEquals(b, null) || a.count != b.count)
                return false;
            for (int i = 0; i < a.count; i++)
            {
                if (!object.Equals(a.items[i], b.items[i]))
                {
                    return false;
                }
            }
            return true;
        }

        public event EventHandler Changed;

        public static bool operator ==(MyList<T> a, MyList<T> b) =>
            Equals(a, b);

        public static bool operator !=(MyList<T> a, MyList<T> b) =>
            !Equals(a, b);
```

File: docs/csharp/tour-of-csharp/program-building-blocks.md [hunk 12]
```diff
@@ -180,70 +125,67 @@ When an instance method declaration includes a `virtual` modifier, the method is
...
 
 Members that contain executable code are collectively known as the *function members* of a class. The preceding section describes methods, which are the primary types of function members. This section describes the other kinds of function members supported by C#: constructors, properties, indexers, events, operators, and finalizers.
 
 The following example shows a generic class called `MyList<T>`, which implements a growable list of objects. The class contains several examples of the most common kinds of function members.
 
-> [!NOTE]
-> This example creates a `MyList` class, which is not the same as the .NET standard <xref:System.Collections.Generic.List%601?displayProperty=nameWithType>. It does illustrate the concepts needed for this tour, but is not a replacement for that class.
-
-[!code-csharp[ListClass](~/samples/snippets/csharp/tour/classes-and-objects/ListBasedExamples.cs#L4-L89)]
+:::code language="csharp" source="./snippets/shared/ClassesObjects.cs" ID="ListExample":::
```
Comment (IEvangelist) [inline docs/csharp/tour-of-csharp/program-building-blocks.md:162]: I think this is the wrong example to show here, it's showing adding to the list but not using indexers.

File: docs/csharp/tour-of-csharp/program-building-blocks.md [hunk 13]
```diff
@@ -255,31 +197,31 @@ An *indexer* is a member that enables objects to be indexed in the same way as a
...
-[!code-csharp[ListExample3](~/samples/snippets/csharp/tour/classes-and-objects/ListBasedExamples.cs#L109-L117)]
+:::code language="csharp" source="./snippets/shared/ClassesObjects.cs" ID="ListAddition":::
 
-Indexers can be overloaded, meaning that a class can declare multiple indexers as long as the number or types of their parameters differ.
+Indexers can be overloaded. A class can declare multiple indexers as long as the number or types of their parameters differ.
 
 ### Events
 
-An *event* is a member that enables a class or object to provide notifications. An event is declared like a field except that the declaration includes an event keyword and the type must be a delegate type.
+An *event* is a member that enables a class or object to provide notifications. An event is declared like a field except that the declaration includes an `event` keyword and the type must be a delegate type.
```
Comment (IEvangelist) [inline docs/csharp/tour-of-csharp/program-building-blocks.md:206]: Curious if we should callout events as supporting multicast delegates?

File: docs/csharp/tour-of-csharp/snippets/shared/ClassesObjects.cs [hunk 14]
```diff
@@ -0,0 +1,453 @@
...
+        public static void Usage()
+        {
+            MyList<string> names = new MyList<string>();
+            names.Changed += new EventHandler(ListChanged);
+            names.Add("Liz");
+            names.Add("Martha");
+            names.Add("Beth");
+            Console.WriteLine(changeCount);		// Outputs "3"
+        }
+    }
```
Comment (IEvangelist) [inline docs/csharp/tour-of-csharp/snippets/shared/ClassesObjects.cs:285]:
```suggestion
    class EventExample
    {
        static int s_changeCount;
        
        static void ListChanged(object sender, EventArgs e)
        {
            s_changeCount++;
        }
        
        public static void Usage()
        {
            var names = new MyList<string>();
            names.Changed += new EventHandler(ListChanged);
            names.Add("Liz");
            names.Add("Martha");
            names.Add("Beth");
            Console.WriteLine(changeCount); // "3"
        }
    }
```

File: docs/csharp/tour-of-csharp/program-building-blocks.md [hunk 15]
```diff
@@ -255,31 +197,31 @@ An *indexer* is a member that enables objects to be indexed in the same way as a
...
 
 ### Operators
 
 An *operator* is a member that defines the meaning of applying a particular expression operator to instances of a class. Three kinds of operators can be defined: unary operators, binary operators, and conversion operators. All operators must be declared as `public` and `static`.
 
-The `MyList<T>` class declares two operators, `operator ==` and `operator !=`, and thus gives new meaning to expressions that apply those operators to `MyList` instances. Specifically, the operators define equality of two `MyList<T>` instances as comparing each of the contained objects using their Equals methods. The following example uses the `==` operator to compare two `MyList<int>` instances.
+The `MyList<T>` class declares two operators, `operator ==` and `operator !=`. These overridden opeators give new meaning to expressions that apply those operators to `MyList` instances. Specifically, the operators define equality of two `MyList<T>` instances as comparing each of the contained objects using their `Equals` methods. The following example uses the `==` operator to compare two `MyList<int>` instances.
```
Comment (IEvangelist) [inline docs/csharp/tour-of-csharp/program-building-blocks.md:222]:
```suggestion
The `MyList<T>` class declares two operators, `operator ==` and `operator !=`. These overridden operators give new meaning to expressions that apply those operators to `MyList` instances. Specifically, the operators define equality of two `MyList<T>` instances as comparing each of the contained objects using their `Equals` methods. The following example uses the `==` operator to compare two `MyList<int>` instances.
```

File: docs/csharp/tour-of-csharp/snippets/shared/Features.cs [hunk 16]
```diff
@@ -0,0 +1,212 @@
...
+        }
+        public static void Main()
+        {
+            double[] a = { 0.0, 0.5, 1.0 };
+            double[] squares = Apply(a, (x) => x * x);
+            double[] sines = Apply(a, Math.Sin);
+            Multiplier m = new Multiplier(2.0);
+            double[] doubles = Apply(a, m.Multiply);
+        }
+    }
```
Comment (IEvangelist) [inline docs/csharp/tour-of-csharp/snippets/shared/Features.cs:38]:
```suggestion
    delegate double Function(double x);
    
    class Multiplier
    {
        double _factor;
        
        public Multiplier(double factor) => _factor = factor;
        
        public double Multiply(double x) => x * _factor;
    }
    
    class DelegateExample
    {
        static double[] Apply(double[] a, Function f)
        {
            var result = new double[a.Length];
            for (int i = 0; i < a.Length; i++) result[i] = f(a[i]);
            return result;
        }
        
        public static void Main()
        {
            double[] a = { 0.0, 0.5, 1.0 };
            double[] squares = Apply(a, (x) => x * x);
            double[] sines = Apply(a, Math.Sin);
            Multiplier m = new Multiplier(2.0);
            double[] doubles = Apply(a, m.Multiply);
        }
    }
```

File: docs/csharp/tour-of-csharp/snippets/shared/Features.cs [hunk 17]
```diff
@@ -0,0 +1,212 @@
...
+            this.url = url;
+        }
+
+        public string Url => url;
+
+        public string Topic
+        {
+            get { return topic; }
+            set { topic = value; }
+        }
```
Comment (IEvangelist) [inline docs/csharp/tour-of-csharp/snippets/shared/Features.cs:57]:
```suggestion
        string _url;        
        string _topic;
        
        public HelpAttribute(string url) => _url = url;

        public string Url => _url;

        public string Topic
        {
            get => _topic;
            set => _topic = value;
        }
```

File: docs/csharp/tour-of-csharp/snippets/shared/Features.cs [hunk 18]
```diff
@@ -0,0 +1,212 @@
...
+            System.Reflection.MethodInfo displayMethod = widgetType.GetMethod(nameof(Widget.Display));
+
+            //Gets every HelpAttribute defined for the Widget.Display method
+            object[] displayMethodAttributes = displayMethod.GetCustomAttributes(typeof(HelpAttribute), false);
+
+            if (displayMethodAttributes.Length > 0)
+            {
+                HelpAttribute attr = (HelpAttribute)displayMethodAttributes[0];
+                Console.WriteLine($"Display method help URL : {attr.Url} - Related topic : {attr.Topic}");
+            }
```
Comment (IEvangelist) [inline docs/csharp/tour-of-csharp/snippets/shared/Features.cs:126]:
```suggestion
            Type widgetType = typeof(Widget);

            object[] widgetClassAttributes = widgetType.GetCustomAttributes(typeof(HelpAttribute), false);

            if (widgetClassAttributes.Length > 0)
            {
                HelpAttribute attr = (HelpAttribute)widgetClassAttributes[0];
                Console.WriteLine($"Widget class help URL : {attr.Url} - Related topic : {attr.Topic}");
            }

            System.Reflection.MethodInfo displayMethod = widgetType.GetMethod(nameof(Widget.Display));

            object[] displayMethodAttributes = displayMethod.GetCustomAttributes(typeof(HelpAttribute), false);

            if (displayMethodAttributes.Length > 0)
            {
                HelpAttribute attr = (HelpAttribute)displayMethodAttributes[0];
                Console.WriteLine($"Display method help URL : {attr.Url} - Related topic : {attr.Topic}");
            }
```

**Review by IEvangelist — APPROVED** [review]
Hi @BillWagner - This was a ton of work, great job! It's so much better with your improvements. I left a lot of nit comments and some code suggestions. It's up to you if you want to take them or not. There were some typos and grammatical suggestions in there too. And even several questions!

Looks good to me, :shipit: when you're ready.

File: docs/csharp/tour-of-csharp/features.md [hunk 19]
```diff
@@ -0,0 +1,118 @@
...
+
+:::code language="csharp" source="./snippets/shared/Features.cs" ID="AsyncExample":::
+
+This small sample shows the major features for asynchronous programming:
+
+- The method declaration includes the `async` modifier.
+- The body of the method `await`s the return of the `GetByteArrayAsync` method.
+- The type specified in the `return` statement matches the type argument in the `TaskL<T>` declaration for the method. (A method that returns a `Task` would use `return` statements without any argument).
```
Comment (tdykstra) [inline docs/csharp/tour-of-csharp/features.md:91]:
```suggestion
- The type specified in the `return` statement matches the type argument in the `Task<T>` declaration for the method. (A method that returns a `Task` would use `return` statements without any argument).
```

File: docs/csharp/tour-of-csharp/features.md [hunk 20]
```diff
@@ -0,0 +1,118 @@
...
+
+The following example declares a `HelpAttribute` attribute that can be placed on program entities to provide links to their associated documentation.
+
+:::code language="csharp" source="./snippets/shared/Features.cs" ID="DefineAttribute":::
+
+All attribute classes derive from the <xref:System.Attribute> base class provided by the standard library. Attributes can be applied by giving their name, along with any arguments, inside square brackets just before the associated declaration. If an attribute’s name ends in `Attribute`, that part of the name can be omitted when the attribute is referenced. For example, the `HelpAttribute` can be used as follows.
```
Comment (tdykstra) [inline docs/csharp/tour-of-csharp/features.md:101]: is "the standard library" the right term here?

**Review by tdykstra — APPROVED** [review]
This is great work, clearly written and comprehensive!  I had mostly minor typo comments but also noted one misplaced snippet.

File: docs/csharp/tour-of-csharp/features.md (hunk 20 above)
Comment (BillWagner) [reply docs/csharp/tour-of-csharp/features.md:101]: That's an artifact from when it was in the C# standard. Updated.

File: docs/csharp/tour-of-csharp/program-building-blocks.md [hunk 21]
```diff
@@ -1,94 +1,39 @@
...
-
-An implicit conversion exists from a class type to any of its base class types. A variable of a class type can reference an instance of that class or an instance of any derived class. For example, given the previous class declarations, a variable of type `Point` can reference either a `Point` or a `Point3D`:
-
-[!code-csharp[Point3DExample](~/samples/snippets/csharp/tour/classes-and-objects/Program.cs#L22-L23)]
+- `public`: Access isn't limited.
+- `protected`: Access is limited to this class or classes derived from this class.
+- `internal`: Access is limited to the current assembly (`.exe` or `.dll`).
+- `protected internal`: Access is limited to the containing class, classes derived from the containing class, or classes within the same assembly.
+- `private`: Access is limited to this class.
+- `private protected`: Access is limited to the containing class or classes derived from the containing type within the same assembly.
```
[the remark this answers is not in the document]
Comment (BillWagner) [reply docs/csharp/tour-of-csharp/program-building-blocks.md:36]: no, that's an artifact of the spec language.

File: docs/csharp/tour-of-csharp/types.md [hunk 22]
```diff
@@ -0,0 +1,108 @@
...
+
+Interfaces may employ ***multiple inheritance***. In the following example, the interface `IComboBox` inherits from both `ITextBox` and `IListBox`.
+
+:::code language="csharp" source="./snippets/shared/Types.cs" ID="FirstInterfaces":::
+
+Classes and structs can implement multiple interfaces. In the following example, the class `EditBox` implements both `IControl` and `IDataBound`.
+
+:::code language="csharp" source="./snippets/shared/Types.cs" ID="ImplementInterfaces":::
+
+When a class or struct implements a particular interface, instances of that class or struct can be implicitly converted to that interface type. For example
```
[the remark this answers is not in the document]
Comment (BillWagner) [reply docs/csharp/tour-of-csharp/types.md:74]: Yes, every explicit conversion is an implicit conversion. I'd rather avoid that in this section.

File: samples/snippets/csharp/roslyn-sdk/Tutorials/MakeConst/MakeConst/AnalyzerReleases.Shipped.md [hunk 23]
```diff
@@ -0,0 +1,3 @@
+﻿; Shipped analyzer releases
```
[the remark this answers is not in the document]
Comment (BillWagner) [reply samples/snippets/csharp/roslyn-sdk/Tutorials/MakeConst/MakeConst/AnalyzerReleases.Shipped.md:1]: build artifact. I removed them.

**Pull request merged**